lib/spack/spack/directives.py

```diff
@@ -932,6 +982,244 @@ def _execute_requires(pkg: Type[spack.package_base.PackageBase]):
+    """
+    return remove_all_directive("requirements")
+
+
+@directive("versions")
+def drop_all_versions():
+    """Removes all versions from a package.
+
+    This is typically used when inheriting from another package if the author
+    desires different versions than those available in the parent package.
+    """
+    return remove_all_directive("versions")
+
+
+class DropDirectiveBase(ABC):
+    name = ""
+
+    def __init__(self, when):
+        self.removal_when = _make_when_spec(when)
+
+    @abstractmethod
+    def add_to_filtered(self, filtered, when, directive_entry):
+        pass
+
+    @abstractmethod
+    def get_directive(self, directive_entry):
+        pass
+
+    def remove(self):
+        removal_when = self.removal_when
+        if not removal_when:
+            return
+        complement_versions = removal_when.versions.complement()
+
+        def _remove(pkg):
+            removal_directive = self.make_directive(pkg)
+            directive_dict = getattr(pkg, self.name)
+            filtered = self.filter_directives(
+                complement_versions, directive_dict, removal_directive, removal_when
+            )
+            directive_dict.clear()
+            directive_dict.update(filtered)
+
+        return _remove
+
+    def filter_directives(
+        self, complement_versions, directive_dict, removal_directive, removal_when
+    ):
+        filtered = {}
+        for when, directives in directive_dict.items():
+            for directive_entry in self.iterate_directives(directives):
+                if self.get_directive(directive_entry) == removal_directive:
+                    if when == removal_when:
+                        continue
+                    elif when.versions.intersects(removal_when.versions):
+                        self.handle_intersection(
+                            when, directive_entry, complement_versions, filtered
+                        )
+                    else:
+                        self.add_to_filtered(filtered, when, directive_entry)
+                else:
+                    self.add_to_filtered(filtered, when, directive_entry)
+        return filtered
+
+    def handle_intersection(self, when, directive_entry, complement_versions, filtered):
+        new_when = when.copy()
+        new_when.versions = when.versions.intersection(complement_versions)
+        if new_when.versions != spack.version.VersionList():
+            self.add_to_filtered(filtered, new_when, directive_entry)
+
+    def iterate_directives(self, directives):
+        return directives
+
+
+class DropConflicts(DropDirectiveBase):
+    name = "conflicts"
+
+    def __init__(self, spec, when):
+        DropDirectiveBase.__init__(self, when)
+        self.make_directive = lambda pkg: spack.spec.Spec(spec)
+
+    def add_to_filtered(self, filtered, when, directive_entry):
+        filtered.setdefault(when, []).append(directive_entry)
```

Comment:
It would be great if we could refactor the `add_to_filtered` methods for each directive type to share code with how the entries are added to the dictionary in the directive itself. It seems fragile to have two different ways of registering a conflict / dependency / etc. in a dictionary implemented separately in two different code paths.